src/late_multicellular_stage/SCALIS.cs

```diff
@@ -0,0 +1,272 @@
+            float deltaTau = Mathf.Abs(aRadius - bRadius);
+
+            if (aRadius > bRadius)
+            {
+                (a, b) = (b, a);
+            }
+
+            // Since i is always 3, we can cut down on some calculations and make static coeff array
+            // var coeff = BinomialExpansion(i - 1);
+
+            for (int k = 0; k < coefficients.Length; k++)
+            {
+                value += coefficients[k] * Mathf.Pow(deltaTau, k) * Mathf.Pow(tau0, i - k - 1)
+                    * Convolution(k, i, a, b, pos);
+            }
+        }
+
+        return value / NormalizationFactor(i, sigma);
+    }
+
+    public Color GetColour(Vector3 pos)
+    {
+        if (Points == null)
+            return Colors.Black;
+
+        if (Points.Length == 1)
+            return Points[0].Colour;
+
+        Color colourSum = Colors.Black;
+
+        float contributionSum = 0.0f;
+
+        for (int j = 0; j < Points.Length; j += 2)
+        {
+            float abDistanceSquared = Points[j].Position.DistanceSquaredTo(Points[j + 1].Position);
+
+            Vector3 linePos = ClosestPoint(pos, Points[j].Position, Points[j + 1].Position);
+
+            float contribution = 1.0f / linePos.DistanceSquaredTo(pos);
+
+            contributionSum += contribution;
+
+            colourSum += Points[j].Colour.Lerp(Points[j + 1].Colour, Points[j].Position.DistanceSquaredTo(linePos)
+                / abDistanceSquared) * contribution;
+        }
+
+        colourSum /= contributionSum;
+
+        colourSum.A = 0.0f;
+
+        return colourSum;
+    }
+
+    private float NormalizationFactor(int i, float sigma)
+    {
+        if (i == 2)
+        {
+            return sigma * sigma * 3.14159265f;
+        }
+
+        if (i == 3)
+        {
+            return sigma * sigma * sigma * 2;
+        }
+
+        if (i < 0)
+        {
+            throw new Exception("i is less that 0");
+        }
+
+        return sigma * sigma * (i - 3) / (i - 2) * NormalizationFactor(i - 2, sigma);
+    }
+
+    private List<int> BinomialExpansion(int n)
+    {
+        var coeff = new List<int>(n + 1);
+        coeff.Add(1);
+        for (int i = 0; i < n; i++)
+        {
+            coeff.Add(coeff[i] * (n - i) / (i + 1));
+        }
+
+        return coeff;
+    }
+
+    private Vector3 ClosestPoint(Vector3 pos, Vector3 a, Vector3 b)
+    {
+        var ab = b - a;
+        var ap = pos - a;
+
+        float cos = ab.Dot(ap) / Mathf.Sqrt(ab.LengthSquared() * ap.LengthSquared());
+
+        if (cos < 0)
+            return a;
+
+        Vector3 linePos = a + ab.Normalized() * cos * ap.Length();
+
+        if ((linePos - a).LengthSquared() > ab.LengthSquared())
+            return b;
+
+        return linePos;
+    }
+
+    private float SquareCutoffValue(Vector3 pos, Vector3 a, Vector3 b)
+    {
+        return (ClosestPoint(pos, a, b) - pos).LengthSquared();
+    }
+
+    private Vector3 VectorFromTo(Vector3 from, Vector3 to)
```

Comment:
This seems a bit weird to have as a separate method so I think this should be just inlined whereever this is needed.